src/vm/database/marf.rs

```diff
@@ -148,47 +206,130 @@ impl <S> KeyValueStorage for &mut MarfedKV <S> where S: KeyValueStorage {
+            }
+        }  
     }
 
-    fn get(&mut self, key: &str) -> Option<String> {
-        self.marf.get(&self.chain_tip, key)
-            .or_else(|e| {
-                match e {
-                    MarfError::NotFoundError => Ok(None),
-                    _ => Err(e)
-                }
-            })
-            .expect("ERROR: Unexpected MARF Failure on GET")
-            .map(|marf_value| {
-                let side_key = marf_value.to_hex();
-                self.side_store.get(&side_key)
-                    .expect(&format!("ERROR: MARF contained value_hash not found in side storage: {}",
-                                     side_key))
-            })
+    pub fn make_contract_hash_key(contract: &QualifiedContractIdentifier) -> String {
+        format!("clarity-contract::{}", contract)
+    }
+
+    /// The contract commitment is the hash of the contract, plus the block height in
+    ///   which the contract was initialized.
+    pub fn make_contract_commitment(&mut self, contract_hash: [u8; 32]) -> String {
+        let block_height = match self.marf {
+            None => 0,
+            Some(ref marf) => marf.get_open_chain_tip_height().expect("Metadata write attempted on unopened MARF")
+        };
+        let cc = ContractCommitment { hash: contract_hash, block_height };
+        cc.serialize()
+    }
+
+    /// This function is used to obtain a committed contract hash, and the block header hash of the block
+    ///   in which the contract was initialized. This data is used to store contract metadata in the side
+    ///   store.
+    pub fn get_contract_hash(&mut self, contract: &QualifiedContractIdentifier) -> Result<(BlockHeaderHash, String)> {
+        let key = MarfedKV::make_contract_hash_key(contract);
+        let contract_commitment = self.get(&key).map(|x| ContractCommitment::deserialize(&x))
+            .ok_or_else(|| { CheckErrors::NoSuchContract(contract.to_string()) })?;
+        let ContractCommitment { block_height, hash: contract_hash } = contract_commitment;
+        let bhh = match self.marf {
+            Some(ref mut marf) => {
+                marf.get_bhh_at_height(&self.chain_tip, block_height)
+                    .expect("Should always be able to map from height to block hash.")
+                    .expect("Should always be able to map from height to block hash.")
+            },
+            None => {
+                TrieFileStorage::block_sentinel()
+            }
+        };
+        Ok((bhh, to_hex(&contract_hash)))
     }
 
-    fn has_entry(&mut self, key: &str) -> bool {
+    pub fn get_metadata(&mut self, contract: &QualifiedContractIdentifier, key: &str) -> Result<Option<String>> {
+        let (bhh, _) = self.get_contract_hash(contract)?;
+        Ok(self.side_store.get_metadata(&bhh, &contract.to_string(), key))
+    }
+
+    pub fn insert_metadata(&mut self, contract: &QualifiedContractIdentifier, key: &str, value: &str) -> bool {
```

Comment:
It looks like no one checks the return value of this method.

Reply:
Yeah -- the bool return value was intended for use by callers that do "does this metadata already exist?" checks before insertion. Currently, those checks are performed directly using `has_metadata`, but that requires an additional MARF read, so in the current implementation, this should panic. However, it'd be better to update the callers to actually just use the boolean result to determine whether or not the metadata already existed (that way, eliminate the extra MARF lookup). Will update.

Reply:
Actually, since the `insert_metadata` function is not called directly during analysis checks, the return value cannot be used for this. I'll just switch this back to panicking.

Reply:
Addressed in 505a1cd